llvm/lib/Target/Mips/MipsAsmPrinter.cpp

```diff
@@ -249,6 +254,55 @@ void MipsAsmPrinter::emitJumpTableDest(MCStreamer &OutStreamer,
   EmitToStreamer(OutStreamer, LoadI);
 }
 
+bool MipsAsmPrinter::tryEmitImm48Fix(MCStreamer &OutStreamer,
+                                     const MachineInstr *MI) {
+  if (!FixImm48)
+    return false;
+
+  if (!Subtarget->hasFixImm48())
+    return false;
+
+  if (MI->getOpcode() != Mips::Li_NM &&
+      MI->getOpcode() != Mips::ADDIU48_NM)
+    return false;
+
+  int LastOp = MI->getNumOperands() - 1;
+  auto Last = MI->getOperand(LastOp);
+
+  if (!Last.isImm())
+    return false;
+
+  int64_t Value = Last.getImm();
+
+  auto ImmValueNeedsImm48Fix = [MI](int64_t Value) -> bool {
+    if ((MI->getOpcode() == Mips::Li_NM) &&
+        (isUInt<16>(Value) || isInt<9>(Value) || ((Value && 0xfffu) == 0u)))
+      return false;
+    return ((Value & (1u << 9)) == (1u << 9)) ||
+           ((Value & (7u << 8)) == (1u << 8)) ||
+           (Value & 0x64005001u) == (0x24005001u);
+  };
+
+  if (!ImmValueNeedsImm48Fix(Value))
+    return false;
```

Comment:
Single use of lambda, would be better to just use directly?

Reply:
Or I guess a ```static``` function for factoring out the logic of the code is even better.